dali/python/nvidia/dali/pool.py

```diff
@@ -0,0 +1,327 @@
+
+        Parameters
+        ----------
+        `context_i` : int
+            Specifies which callback will be used to run the task, it must be the index corresponding
+            to the order of callbacks passed when constructing WorkerPool.
+        `batch_i` : int
+            Ordinal of the batch that tasks list corresponds to.
+        `tasks` : nvidia.dali.types.SampleInfo list
+            You can think of resulting batch as [callback(task) for task in tasks] with the exception that
+            callbacks will be run in parallel.
+        """
+        tasks = list(enumerate(tasks))
+        if not tasks:
+            raise RuntimeError("Cannot schedule empty list of tasks")
+        context = self.contexts[context_i]
+        if batch_i in context.scheduled:
+            raise RuntimeError("Given batch has already been scheduled")
+        self._distribute((context_i, batch_i), tasks)
+        context.partially_received[batch_i] = {}
+        context.scheduled[batch_i] = tasks
+
+    def _distribute(self, header, tasks):
+        workers_num = self.pool.workers_num
+        tasks_no = len(tasks)
+        chunk_size = tasks_no // workers_num
+        queued_no = chunk_size + (tasks_no % workers_num)
+        with self.pool.task_pipes_lock:
+            task_pipes = self.pool.task_pipes
+            task_pipes[0].send((header, tasks[:queued_no]))
+            if chunk_size:
+                for worker_id in range(1, workers_num):
+                    task_pipes[worker_id].send((header, tasks[queued_no:queued_no + chunk_size]))
+                    queued_no += chunk_size
+
+    def receive_batch(self, context_i):
+        """Returns the next resulting batch (in the order of schedule_batch calls) for the ``context_i``th callaback.
+
+        Parameters
+        ----------
+        `context_i` : int
+            Specifies which callback you want the results from, ordering corresponds to the order of
+            callbacks passed when constructing the pool.
+        """
+        context = self.contexts[context_i]
+        batch_i, tasks = context.scheduled.popitem(last=False)
+        awaiting_batch = context.partially_received[batch_i]
+        while len(awaiting_batch) < len(tasks) and batch_i not in context.iter_ended:
+            self._receive_chunk()
+        if batch_i in context.iter_ended:
+            context.iter_ended.remove(batch_i)
+            del context.partially_received[batch_i]
+            raise StopIteration
+        res = [awaiting_batch[i] for i, _ in tasks]
+        del context.partially_received[batch_i]
+        return res
+
+    def _receive_chunk(self):
+        ready_workers = multiprocessing.connection.wait(self.rec_pipes)
+        for worker_pipe in ready_workers:
+            message = worker_pipe.recv()
+            if message is None:
+                raise RuntimeError("Worker exited unexpectedly")
+            worker_id, header, worker_batch_serialized = message
```

Comment:
Can we make the massage a class that has a well defined fields and we can keep easily in sync between pool and the worker?

Reply:
Made them into a class